File: Python/src/gestorAplicacion/clasesBase/OfertaPorDefecto.py
from typing import List
from Producto import *
from Inventariar import *
from Producto import *


class OfertaPorDefecto(Inventariar):
    def __init__(self, productosVendidos):
        self.productosOferta = self.establecerOferta(productosVendidos)
        self.total = sum(producto.precio for producto in self.productosOferta)

    def establecerOferta(self, productosVendidos):
        contadores = [0, 0, 0]
        tipoMenosVendido = None
        listaAuxiliar = []

        contadores[0] = self.calcularCamisas(productosVendidos)
        contadores[1] = self.calcularAbrigos(productosVendidos)
        contadores[2] = self.calcularPantalon(productosVendidos)
        minContador = min(contadores)

        if contadores[0] <= minContador:
            tipoMenosVendido = Producto.Tipo.CAMISA
        if contadores[1] <= minContador:
            tipoMenosVendido = Producto.Tipo.ABRIGO
        if contadores[2] <= minContador:
            tipoMenosVendido = Producto.Tipo.PANTALON

        productoParaAgregar = Producto(tipoMenosVendido, 0.9 * tipoMenosVendido.value, 15000)

        for _ in range(9):
            if not(tipoMenosVendido.__eq__(Producto.Tipo.ABRIGO)) and self.calcularAbrigos(lista_auxiliar) < 2:
                lista_auxiliar.append(Producto(Producto.Tipo.ABRIGO, 0.9 * Producto.Tipo.ABRIGO.value(), 15000))
            elif not(tipoMenosVendido.__eq__(Producto.Tipo.CAMISA)) and self.calcular_camisas(lista_auxiliar) < 2:
                lista_auxiliar.append(Producto(Producto.Tipo.CAMISA, 0.9 * Producto.Tipo.CAMISA.value(), 15000))
            elif not(tipoMenosVendido.__eq__(Producto.Tipo.PANTALON)) and self.calcular_pantalon(lista_auxiliar) < 2:
                listaAuxiliar.append(Producto(Producto.Tipo.PANTALON, 0.9 * Producto.Tipo.PANTALON.value(), 15000))
            elif listaAuxiliar.count(productoParaAgregar) < 5:
                listaAuxiliar.append(productoParaAgregar)

        return listaAuxiliar

    def getProductosOferta(self):
        return self.productosOferta

    def setProductosOferta(self, productosOferta):
        self.productosOferta=productosOferta

    def getTotal(self):
        return self.total

    def setTotal(self, total):
        self.total=total
    def __lt__(self, otra):
        if(self.total < otra.getTotal()):
            return True
        else:
            return False


    def __str__(self):
        return f"OfertaPorDefecto: productos={self.productosOferta}, total={self.total}"

File: AdministracionDeTienda/src/gestorAplicacion/clasesBase/OfertaPorDefecto.java
package gestorAplicacion.clasesBase;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import gestorAplicacion.clasesBase.Producto.Tipo;

public class OfertaPorDefecto implements Comparable<OfertaPorDefecto>,Serializable{


	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;
	private ArrayList<Producto> productosOferta;
	private double total=0;


	public OfertaPorDefecto(ArrayList<Producto> productosVendidos) {
		productosOferta=establecerOferta(productosVendidos);
		for(Producto p: productosOferta) {
			total+=p.getPrecio();
		}


	}

	/*
	protected OfertaPorDefecto(int i) {
		this.i=ponerNumero(i);
		System.out.println(this.i);

	}

	public int ponerNumero(int i) {
		return i++;
	}
	*/
	public ArrayList<Producto> establecerOferta(ArrayList<Producto> productosVendidos) {
		Integer[] contadores= {0,0,0};

		Tipo tipoMenosVendido = null;
		ArrayList<Producto> listaAuxiliar= new ArrayList<Producto>();
		contadores[0]=Inventariar.calcularCamisas(productosVendidos);
		contadores[1]=Inventariar.calcularAbrigos(productosVendidos);
		contadores[2]=Inventariar.calcularPantalon(productosVendidos);
		List<Integer> ints = Arrays.asList(contadores);
		int min=Collections.min(ints);


		if(contadores[0]<=min){
				tipoMenosVendido=Tipo.CAMISA;
		}
		if(contadores[1]<=min) {
			tipoMenosVendido=Tipo.ABRIGO;
		}
		if(contadores[2]<=min) {
			tipoMenosVendido=Tipo.PANTALON;
		}

		Producto productoParaAgregar=new Producto(tipoMenosVendido,0.9f*tipoMenosVendido.getPrecioEstandar(),15000);


		for(int i=0; i<9;i++) {
		//Relleno la lista

			if(!(tipoMenosVendido.equals(Tipo.ABRIGO)) && Inventariar.calcularAbrigos(listaAuxiliar)<2 ) {

				listaAuxiliar.add(new Producto(Tipo.ABRIGO,0.9f*Tipo.ABRIGO.getPrecioEstandar(),15000));

			}
			else if(!(tipoMenosVendido.equals(Tipo.CAMISA)) && Inventariar.calcularCamisas(listaAuxiliar)<2) {
				listaAuxiliar.add(new Producto(Tipo.CAMISA,0.9f*Tipo.CAMISA.getPrecioEstandar(),15000));

			}
			else if(!(tipoMenosVendido.equals(Tipo.PANTALON)) && Inventariar.calcularPantalon(listaAuxiliar)<2) {
				listaAuxiliar.add(new Producto(Tipo.PANTALON,0.9f*Tipo.PANTALON.getPrecioEstandar(),15000));

			}
			else if(Collections.frequency(listaAuxiliar, productoParaAgregar)<5) {
				listaAuxiliar.add(productoParaAgregar);

			}

		}

		return listaAuxiliar;
	}


	public ArrayList<Producto> getProductosOferta() {
		return productosOferta;
	}

	public void setProdutosOferta(ArrayList<Producto> productosOferta){
		this.productosOferta=productosOferta;
	}

	public double getTotal() {
		return total;
	}

	public void setTotal(double total) {
		this.total = total;
	}
	//La mejor oferta es aquella con menor total
		//Yo vengo antes que la otra si es NEGATIVO---ORDENA DE MENOR A MAYOR

	@Override
	public int compareTo(OfertaPorDefecto otra) {
		return (int)(this.total-otra.getTotal());
	}


}
